Cython/Compiler/ExprNodes.py

```diff
@@ -5026,23 +5029,59 @@ def coerce_to(self, dst_type, env):
                     "default encoding required for conversion from '%s' to '%s'" %
                     (self.base.type, dst_type))
             self.type = dst_type
-        if dst_type.is_array and self.base.type.is_array:
-            if not self.start and not self.stop:
-                # redundant slice building, copy C arrays directly
-                return self.base.coerce_to(dst_type, env)
-            # else: check array size if possible
+        #elif (self.base.type.is_array or self.base.type.is_ptr) and dst_type.is_pyobject:
+        elif self.base.type.is_array:
+            if dst_type.is_pyobject:
+                if dst_type is tuple_type:
+                    to_py_obj_type = 'tuple'
+                elif dst_type is set_type:
+                    to_py_obj_type = 'set'
+                elif dst_type is frozenset_type:
+                    to_py_obj_type = 'frozenset'
+                elif dst_type.is_pyobject:
+                    to_py_obj_type = 'list'
+                else:
+                    error(self.pos,
+                          "Cannot convert slice of C array to '%s'" % dst_type)
+
+                self.type = dst_type
+                to_py_func_cname = "__Pyx_slice_carray_" + self.base.type.specialization_name()\
+                                        + '_to_py_' + to_py_obj_type
```

Comment:
I think we normally prefer splitting lines with brackets than `\`, and string formatting to concatenation:

```
to_py_func_cname = "__Pyx_slice_carray_%s_to_py_%s" % (
      self.base.type.specialization_name(), to_py_obj_type)
```

Although don't copy my indenting because I never get it right